build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -0,0 +1,135 @@
+    if (!path.get('body').isBlockStatement()) {
+      return false;
+    }
+    const body = path.get('body.body.0');
+    return !body || !body.isReturnStatement();
+  },
+
+  // Since we don't know if exported members are 'new'd, bail out on modification.
+  isExported: (t, path) =>
+    path.findParent(
+      path =>
+        path.isExportNamedDeclaration() || path.isExportDefaultDeclaration()
+    ),
+
+  // If the function contains usage of `arguments`, bail out on modification.
+  containsArgumentsUsage: (t, path) => {
+    let containsArgumentsUsage = false;
+    path.traverse({
+      Identifier(path) {
+        if (t.isIdentifier(path.node, {name: 'arguments'})) {
+          containsArgumentsUsage = true;
+          path.stop();
+        }
+      },
+    });
+    return containsArgumentsUsage;
+  },
+
+  // If the function contains a ThisExpression, converting from a FunctionDeclaration to a
+  // VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+  // Bail out on modification.
+  containsThisExpression: (t, path) => {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  },
+
+  // If the FunctionDeclaration identifier is newed in the scope of this program, bail out on modification.
+  isNewedInProgramScope: (t, path) => {
+    const {name} = path.get('id').node;
+    let isNewed = false;
+    path
+      .findParent(path => path.isProgram())
+      .traverse({
+        NewExpression(path) {
+          if (t.isIdentifier(path.node.callee, {name})) {
+            isNewed = true;
+            path.stop();
+          }
+        },
+      });
+    return isNewed;
+  },
+};
+
+function createVariableDeclaration(t, path) {
+  const {params, body, async, id} = t.cloneNode(path.node);
+  const arrowFunction = t.arrowFunctionExpression(
+    params,
+    body.body[0].argument,
+    async
+  );
+  const declarations = t.variableDeclarator(
+    t.identifier(id.name),
+    arrowFunction
+  );
+
+  // Making an existing FunctionDeclaration a const VariableDeclaration is safe.
+  // The name was already reserved for the FunctionDeclaration.
+  return t.variableDeclaration('const', [declarations]);
+}
+
+// Attempt to convert simple single ReturnStatement FunctionDeclarations to ArrowFunctionExpressions.
+// See BAIL_OUT_CONDITIONS for reasons why FunctionDeclarations would not be modified.
+export default function({types: t}) {
+  const DEBUG = true;
+
+  return {
+    name: 'safe-arrows',
+    visitor: {
+      FunctionDeclaration(path) {
+        for (const [id, method] of Object.entries(BAIL_OUT_CONDITIONS)) {
+          if (method(t, path)) {
+            if (DEBUG) {
+              console /*OK*/
+                .log(
+                  `Bail on ${
+                    path.get('id') && path.get('id').node.name
+                      ? `function ${path.node.id.name}:`
```

Comment:
Could we add a message string function in conjunction with "avoid else, return/continue early" to prevent [Hadouken nesting](https://i.imgur.com/BtjZedW.jpg)? 😬

Reply:
Sure thing, makes sense.

Reply:
Lol @ hadouken nesting.